File: api/app/routers/sources.py
"""Add reference documents to the repository."""
from fastapi import APIRouter, Depends, UploadFile, File, HTTPException
from sqlalchemy.orm import Session
from app.db.session import get_db
from app.services.extract import extract_text, UnsupportedFile
from app.services.similarity.index import add_document
from app.models.document import DocKind

router = APIRouter(prefix="/api/v1", tags=["sources"])

@router.post("/sources")
async def add_source(file: UploadFile = File(...), db: Session = Depends(get_db)):
    try:
        text = extract_text(file.filename, await file.read())
    except UnsupportedFile as e:
        raise HTTPException(415, str(e))
    doc = add_document(db, file.filename, text, DocKind.source)
    return {"id": doc.id, "title": doc.title, "word_count": doc.word_count}

File: web/src/api/sources.ts
import { api } from "./client";
import type { SourceDoc } from "./types";
export const listSources = () => api.get<SourceDoc[]>("/sources").catch(() => [] as SourceDoc[]);
export function addSource(file: File) {
  const fd = new FormData(); fd.append("file", file);
  return api.post<SourceDoc>("/sources", fd);
}
